solls/LanguageServer.cpp

```diff
@@ -0,0 +1,679 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <solls/LanguageServer.h>
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTVisitor.h>
+#include <libsolidity/interface/ReadFile.h>
+
+#include <liblangutil/SourceReferenceExtractor.h>
+
+#include <libsolutil/Visitor.h>
+#include <libsolutil/JSON.h>
+
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+
+#include <ostream>
+
+#include <iostream>
+#include <string>
+
+using namespace std;
+using namespace std::placeholders;
+
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+
+namespace solidity {
+
+namespace // {{{ helpers
+{
+
+// TOOD: maybe use SimpleASTVisitor here, if that would be a simple free-fuunction :)
+class ASTNodeLocator : public ASTConstVisitor
+{
+private:
+	int m_pos = -1;
+	ASTNode const* m_currentNode = nullptr;
+
+public:
+	explicit ASTNodeLocator(int _pos): m_pos{_pos}
+	{
+	}
+
+	ASTNode const* closestMatch() const noexcept { return m_currentNode; }
+
+	bool visitNode(ASTNode const& _node) override
+	{
+		if (_node.location().start <= m_pos && m_pos <= _node.location().end)
+		{
+			m_currentNode = &_node;
+			return true;
+		}
+		return false;
+	}
+};
+
+// TOOD: maybe use SimpleASTVisitor here, if that would be a simple free-fuunction :)
+class ReferenceCollector: public frontend::ASTConstVisitor
+{
+private:
+	frontend::Declaration const& m_declaration;
+	std::vector<lsp::Server::DocumentHighlight> m_result;
+
+public:
+	explicit ReferenceCollector(frontend::Declaration const& _declaration):
+		m_declaration{_declaration}
+	{
+		fprintf(stderr, "finding refs: '%s', '%s'\n",
+				_declaration.name().c_str(),
+				_declaration.location().text().c_str());
+	}
+
+	std::vector<lsp::Server::DocumentHighlight> take() { return std::move(m_result); }
+
+	bool visit(Identifier const& _identifier) override
+	{
+		if (auto const declaration = _identifier.annotation().referencedDeclaration; declaration)
+			if (declaration == &m_declaration)
+				addReference(_identifier.location(), "(identifier)");
+
+		return visitNode(_identifier);
+	}
+
+	void addReference(SourceLocation const& _location, char const* msg = "")
+	{
+		auto const [startLine, startColumn] = _location.source->translatePositionToLineColumn(_location.start);
+		auto const [endLine, endColumn] = _location.source->translatePositionToLineColumn(_location.end);
+		auto const locationRange = lsp::Range{
+			lsp::Position{startLine, startColumn},
+			lsp::Position{endLine, endColumn}
+		};
+
+		fprintf(stderr, " -> found reference %s at %d:%d .. %d:%d\n",
+			msg,
+			startLine, startColumn,
+			endLine, endColumn
+		);
+
+		auto highlight = lsp::Server::DocumentHighlight{};
+		highlight.range = locationRange;
+		highlight.kind = lsp::Server::DocumentHighlightKind::Text; // TODO: are you being read or written to?
+
+		m_result.emplace_back(highlight);
+	}
+
+	// TODO: MemberAccess
```

Comment:
And `IdentifierPath` - basically look through ASTAnnotations.h and search for all occurrences of referencedDeclaration.